velox/connectors/hive/storage_adapters/s3fs/RegisterS3FileSystem.cpp

```diff
@@ -26,73 +26,89 @@
-    const velox::dwio::common::FileSink::Options& options)>
-s3WriteFileSinkGenerator() {
-  static auto s3WriteFileSink =
-      [](const std::string& fileURI,
-         const velox::dwio::common::FileSink::Options& options)
-      -> std::unique_ptr<dwio::common::WriteFileSink> {
-    if (isS3File(fileURI)) {
-      auto fileSystem =
-          filesystems::getFileSystem(fileURI, options.connectorProperties);
-      return std::make_unique<dwio::common::WriteFileSink>(
-          fileSystem->openFileForWrite(fileURI, {{}, options.pool}),
-          fileURI,
-          options.metricLogger,
-          options.stats);
-    }
-    return nullptr;
-  };
+std::shared_ptr<FileSystem> fileSystemGenerator(
+    std::shared_ptr<const Config> properties,
+    std::string_view /*filePath*/) {
+  std::shared_ptr<core::MemConfig> config = std::make_shared<core::MemConfig>();
+  if (properties) {
+    *config = core::MemConfig(properties->values());
+  }
+  const auto s3Identity = getS3Identity(config);
+
+  return fileSystems().withWLock(
+      [&](auto& instanceMap) -> std::shared_ptr<FileSystem> {
+        initializeS3(config.get());
+        if (instanceMap.count(s3Identity) == 0) {
+          auto fs = std::make_shared<S3FileSystem>(properties);
+          instanceMap.insert({s3Identity, fs});
+          return fs;
+        }
+        return instanceMap[s3Identity];
+      });
+}
 
-  return s3WriteFileSink;
+std::unique_ptr<velox::dwio::common::FileSink> s3WriteFileSinkGenerator(
+    const std::string& fileURI,
+    const velox::dwio::common::FileSink::Options& options) {
+  if (isS3File(fileURI)) {
+    auto fileSystem =
+        filesystems::getFileSystem(fileURI, options.connectorProperties);
+    return std::make_unique<dwio::common::WriteFileSink>(
+        fileSystem->openFileForWrite(fileURI, {{}, options.pool}),
+        fileURI,
+        options.metricLogger,
+        options.stats);
+  }
+  return nullptr;
 }
 #endif
 
 void registerS3FileSystem() {
 #ifdef VELOX_ENABLE_S3
-  if (!s3fs) {
-    registerFileSystem(isS3File, fileSystemGenerator());
-    dwio::common::FileSink::registerFactory(s3WriteFileSinkGenerator());
-  }
+  fileSystems().withWLock([&](auto& instanceMap) {
+    if (instanceMap.empty()) {
+      registerFileSystem(isS3File, std::function(fileSystemGenerator));
+      dwio::common::FileSink::registerFactory(
+          std::function(s3WriteFileSinkGenerator));
+    }
+  });
 #endif
 }
 
 void finalizeS3FileSystem() {
 #ifdef VELOX_ENABLE_S3
-  VELOX_CHECK(
-      !s3fs || (s3fs && s3fs.use_count() == 1),
-      "Cannot finalize S3FileSystem while in use");
-  s3fs.reset();
+  bool singleUseCount = true;
+  fileSystems().withWLock([&](auto& instanceMap) {
+    for (const auto& [id, fs] : instanceMap) {
+      singleUseCount &= (fs.use_count() == 1);
+    }
+  });
+
+  VELOX_CHECK(singleUseCount, "Cannot finalize S3FileSystem while in use");
+  fileSystems().withWLock([&](auto& instanceMap) { instanceMap.clear(); });
```

Comment:
feels like you should do that within the same lock to prevent a race where someone just reacquires a pointer in between.